theano/gpuarray/opt.py

```diff
@@ -210,6 +210,52 @@ def apply(self, fgraph):
                     0, 'fast_run', 'fast_compile', 'merge')
 
 
+class GraphToGPU(Optimizer):
+    """
+    Transfer the graph as a whole to GPU instead of replacing nodes
+    """
+    def add_requirements(self, fgraph):
+        fgraph.attach_feature(toolbox.ReplaceValidate())
+
+    def apply(self, fgraph):
+        mapping = {}
+        move_to_GPU = True
+
+        # Building a new graph
+        for i in fgraph.inputs:
+            mapping[i] = GpuFromHost(None)(i)
+
+        for node in fgraph.toposort():
+
+            # The Extra condition
+            if node.inputs is node.outputs:
+                move_to_GPU = False
+
+            # Oplifter's condition
+            # Will return a list of OP
+            # If None, means can't be moved.
+
+            new_ops = local_gpuaalloc(node, None)
+
+            if new_ops is None:
+                move_to_GPU = False
+
+            if not isinstance(new_ops[0], node.op):
+                move_to_GPU = False
+                continue
+
+            if move_to_GPU:
+                for old_o, new_o in zip(node.outputs, new_ops):
+                    mapping[old_o] = new_o
+            else:
+                for o in node.outputs:
+                    mapping[o] = o
+
```

Comment:
For your first question, yes and no. If you replace a node, it will replace all the previous nodes/variables to that node. We want this to be done for all the graph outputs. Then this will change the full graph. If we do it in separate call, and the the previous nodes from 2 outputs nodes are shared, then there can be problems. If 2 outputs nodes shared previous nodes and we do the change in 1 call to replace_all_validate, then all should be fine.